File: solutions/java/all-your-base/1/src/main/java/BaseConverter.java
import java.util.List;
import java.util.ArrayList;


class BaseConverter {

    int myValue = 0;

    BaseConverter(int originalBase, int[] originalDigits) {
        if(originalBase<2){
           throw new IllegalArgumentException("Bases must be at least 2.");
        }
        int myBase = originalBase;
        int power = 1; 
        for(int i=originalDigits.length-1; i>=0; i--){
            int d=originalDigits[i];
            if(d<0) {
                throw new IllegalArgumentException("Digits may not be negative.");
            }
            else if(d>=myBase){
                 throw new IllegalArgumentException("All digits must be strictly less than the base.");
            }
            myValue += originalDigits[i]*power;
            power *= myBase;
        }
    }

    int[] convertToBase(int newBase) {
       if(newBase<2){
           throw new IllegalArgumentException("Bases must be at least 2.");
        }
        if(0==myValue)
            return new int[]{0};
        int num = myValue;
        List<Integer> newDigits = new ArrayList<>();
        while (num>0){
            newDigits.addFirst(num % newBase);
            num /= newBase;
        }
        int[] digits = new int[newDigits.size()];
        for(int i=0; i<digits.length; i++){
            digits[i] = newDigits.get(i);
        }
        return digits;
    }

}

File: solutions/kotlin/all-your-base/1/src/main/kotlin/BaseConverter.kt
class BaseConverter(val base: Int, val digits: IntArray) {

    init{
        require(base >=2){"Bases must be at least 2."}
        require(digits.isNotEmpty()){"You must supply at least one digit."}
        if (digits.size>1){
            require(digits[0]!=0){"Digits may not contain leading zeros."}
        }
        require(digits.all{ it>=0}){"Digits may not be negative."}
        require(digits.all{it<base}){"All digits must be strictly less than the base."}
    }
    
   
    
    
    fun convertToBase(newBase: Int): IntArray {
        require(newBase >=2){"Bases must be at least 2."}
        var n = 0;
        for (d in digits){
            n = base*n + d
        }
        if (n==0) return IntArray(1){0}
        
        var output = mutableListOf<Int>()
        while (n>0) {
            output.add(n % newBase)
            n /= newBase
        }
        return output.asReversed().toIntArray()
    }
}
